src/libpmemfile-posix-fake/pmemfilepool.c

```diff
@@ -0,0 +1,1474 @@
+	mode_t previous_umask_value = umask(0);
+	// pmemfile creates directories as 0777, so we need to replicate that
+	int result = mkdir(newpath, 0777);
+	umask(previous_umask_value);
+	if (result == 0 || (result != 0 && errno == EEXIST)) {
+		poolsize_get_set(pathname, &poolsize, true);
+		chdir(newpath);
+
+		PMEMfilepool *ret = malloc(sizeof(PMEMfilepool));
+		pmemfilepool_init(ret, newpath, poolsize);
+		free(newpath);
+
+		return ret;
+	}
+	free(newpath);
+	return NULL;
+}
+
+int
+pmemfile_getdents64(PMEMfilepool *pfp, PMEMfile *file,
+			struct linux_dirent64 *dirp, unsigned count)
+{
+	if (pfp == NULL || file == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	return (int)syscall(SYS_getdents64, file->fd, dirp, count);
+}
+
+int
+pmemfile_fstatat(PMEMfilepool *pfp, PMEMfile *dir, const char *path,
+		struct stat *buf, int flags)
+{
+	if (pfp == NULL || dir == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	/*
+	 * Memcheck reports error when calling POSIX function
+	 * with NULL-valued parameters, however POSIX function
+	 * will set error code and return -1
+	 * Syscall param fstatat(file_name) points to unaddressable byte(s)
+	 */
+	if (on_valgrind && (path == NULL || buf == NULL)) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	int result = fstatat(dir->fd, path, buf, flags);
+
+	/*
+	 * Symlink size is length of path it points to, so we need
+	 * to subtract length of pool path from it
+	 * hack for tests/posix/stat/stat.cpp - fstatat test
+	 */
+	if (result != -1 && S_ISLNK(buf->st_mode)) {
+		buf->st_size -= (ssize_t)strlen(pfp->pool_path);
+		if (buf->st_size < 0)
+			buf->st_size = 0;
+	}
+
+	return result;
+}
+
+int
+pmemfile_unlink(PMEMfilepool *pfp, const char *pathname)
+{
+	if (pfp == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	if (pathname == NULL) {
+		errno = ENOENT;
+		return -1;
+	}
+
+	char *full_path = merge_paths(pfp->pool_path, pathname);
+	int result = unlink(full_path);
+
+	free(full_path);
+
+	return result;
+}
+
+int
+pmemfile_link(PMEMfilepool *pfp, const char *oldpath, const char *newpath)
+{
+	if (pfp == NULL) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	if (oldpath == NULL || newpath == NULL) {
+		errno = ENOENT;
+		return -1;
+	}
+
+	char *full_old_path = merge_paths(pfp->pool_path, oldpath);
+	char *full_new_path = merge_paths(pfp->pool_path, newpath);
+
+	int result = link(full_old_path, full_new_path);
+
+	free(full_old_path);
+	free(full_new_path);
+
+	return result;
+}
+
+void
+pmemfile_stats(PMEMfilepool *pfp, struct pmemfile_stats *stats)
+{
+	stats->blocks = 0;
+	stats->block_arrays = 0;
+	stats->dirs = 0;
+	stats->inodes = 0;
+	stats->inode_arrays = 0;
+}
+
+int
+pmemfile_mkdir(PMEMfilepool *pfp, const char *path, mode_t mode)
+{
+	if (pfp == NULL || path == NULL) {
+		errno = ENOENT;
+		return -1;
+	}
+
+	char *full_path = merge_paths(pfp->pool_path, path);
+
+	/*
+	 * libpmemfile-posix does not support umask
+	 * hack for any test using pmemfile_mkdir
+	 */
+	mode_t previous_umask = umask(0);
```

Comment:
I think this should be fixed by calling pmemfile_umask(pfp, 0) from pmemfile_test::SetUp